File: 2415-reverse-odd-levels-of-binary-tree/2415-reverse-odd-levels-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def reverseOddLevels(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root:
            return None     # Return None if the tree is empty.
        
        queue, level = deque([root]), 0     # Start BFS with the root node
        
        while queue:
            n, curr_lvl_nodes = len(queue), []
            
            # Process all nodes at the current level.
            for _ in range(n):
                node = queue.popleft()
                curr_lvl_nodes.append(node)
                
                if node.left:
                    queue.append(node.left)
                
                if node.right:
                    queue.append(node.right)
                
            # Reverse node values if the current level is odd
            if level % 2:
                left, right = 0, len(curr_lvl_nodes) - 1
                
                while left < right:
                    curr_lvl_nodes[left].val, curr_lvl_nodes[right].val = curr_lvl_nodes[right].val, curr_lvl_nodes[left].val
                    left += 1
                    right -= 1
            
            level += 1
            
        return root     # Return the modified tree root.

File: 2415-reverse-odd-levels-of-binary-tree/2415-reverse-odd-levels-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* reverseOddLevels(TreeNode* root) {
        if (!root) return root;     // Return null if the tree is empty.
        
        int level = 0;
        queue<TreeNode*> queue;
        queue.push(root);   // Start BFS with the root node.
        
        while (!queue.empty()){
            int n = queue.size();
            vector<TreeNode*> currentLevelNodes;
            
            // Process all nodes at the current level.
            for (int i = 0; i < n; ++i){
                TreeNode* node = queue.front(); queue.pop();
                currentLevelNodes.push_back(node);
                
                if (node -> left)
                    queue.push(node -> left);
                
                if (node -> right)
                    queue.push(node -> right);
            }
            
            // Reverse node values if the current level is odd.
            if (level % 2){
                int left = 0, right = currentLevelNodes.size() - 1;
                while (left < right){
                    swap(currentLevelNodes[left++] -> val, currentLevelNodes[right--] -> val);
                }
            }
            
            ++level;
        }
        
        return root;    // Return the modified tree root.
    }
};